### ISSUE TITLE
Local printers keep disappearing

### ISSUE THREAD

**Issue by kylemacfarlane** [google/cups-connector#200]
I have two printers shared via a Raspberry Pi print server. They are always available through cloud print but the local print versions keep disappearing on Windows and Chrome OS (so far Android has always seen them but that might just be luck). Sometimes they are only gone for a minute or two but other times they are gone until I restart gcp or avahi-daemon (restarting either will fix it).

I have logging set to debug level and when the local printers disappear it seems like the connector stops receiving the privet requests to /capabilities, /info, etc.

So I installed Bonjour Browser on Android and on Windows and Chrome OS I installed this Chrome app: https://chrome.google.com/webstore/detail/mdns-browser/kipighjpklofchgbdgclfaoccdlghidp Even when the local printers aren't appearing in Chrome both these browsers can find the printers on privet.

Restarting the gcp service seems to be a 100% fix to make the local printers instantly reappear so presumably whatever it does on startup needs to be done more frequently.

(Also Chrome on iOS can never see the local printers but I assume that's just an iOS limitation.)

**Comment by kylemacfarlane** [comment]
To make matters worse, if you click the print button and the local printer happens to disappear in the next second or so then Chrome will actually send the print job to the OS default printer and not the printer you  selected. You can  see the selected printer flash to another printer just before the UI closes.

**Comment by jacobmarble** [comment]
I assume that "gcp service" means this project, the CUPS Connector.

What do you mean that they disappear on Windows? Chrome browser running on Windows?

Thanks for following up with checking the mDNS status from other applications. Unfortunately, we rely on Avahi to manage mDNS "advertising" so there's not a lot we can do to fix that. It could be the ChromeOS implementation of mDNS/Cloud Print client, which I'm not familiar with.

Feel free to dig into the Avahi client code inside the connector, see if there's a flag or some argument we can tweak. If you identify something, I'll be happy to make the code change (not sure what your programming abilities are).

You are right, the iOS Cloud Print implementation doesn't support Privet.

**Comment by kylemacfarlane** [comment]
Sorry, I forgot that the gcp.service actually came from the wiki rather than this project itself. But yeah, when I say restart I mean stop and start gcp-cups-connector.

Yes for Windows I mean it disappears from the list of printers in the Chrome browser just like it does on Chrome OS. The Chrome browser on Windows and Chrome OS are both displaying identical problems. The Chrome browser on Android hasn't had any problems yet but it seems like Android detects all kinds of network printers by default so clearly it does something different.

Another issue is that sometimes Chrome can't seem to find the last used local printer (it shows the spinner next to the printer under "Destination") and then reverts to the OS default. But then when I open the UI to select a different printer the local printer is in there albeit not under "Recent Destinations". So I don't know if some underlying ID for the printer is changing and Chrome thinks it's seeing a new printer.

**Comment by jacobmarble** [comment]
That sounds like a Chrome bug; you should report it at crbug.com. Same for the recent printers problem.

**Comment by kylemacfarlane** [comment]
Using Wireshark this looks more and more like a Chrome bug but it should be easily repeatable so I wonder if someone else can check?

What I'm seeing is that Chrome sends privet requests before mDNS responses are received. So if a local printer is in the mDNS cache it will work fine but if the cache has expired then Chrome will never query the printer for privet info.

1. Use some kind of network monitoring tool.

2. Open a print dialog in Chrome and then go to change printers and select your locally shared printer. This should fill your cache and get you in the same starting position.

3. But don't print anything. Just close that print dialog and refresh the page.

4. Open another print dialog. You should see the following traffic:
4.1. One or more mDNS requests
4.2. A HTTP privet request
4.3. A HTTP privet response
4.4. One or more mDNS responses

5. Again don't print anything. Close the dialog, refresh the page, and wait ten or so minutes for Chrome to clear it's cache (clearing the main net-internals DNS cache doesn't seem to affect the mDNS cache).

6. Open a new print dialog. You know the cache is empty when you see the loading spinner next to your printer and no printer options below it. You should see the following traffic:
6.1. One or more mDNS requests
6.2. No HTTP requests (the mDNS cache is presumably empty after step 5).
6.3. One or more mDNS responses

7. Click change printer. You should see the following traffic:
7.1. One or more mDNS requests
7.2. A HTTP privet request (the mDNS cache is presumably primed after step 6.3)
7.3. A HTTP privet response
7.4. One or more mDNS responses


If the above is a real bug then surely it would affect everyone using GCP 2.0? However I also noticed that the mDNS privet responses include some cache settings so maybe they are breaking Chrome?

**Comment by jacobmarble** [comment]
Kyle, you're doing a lot of work here, documenting the bug very well. Be certain to copy-paste to a new bug at crbug.com and then drop a link here.

**Comment by agoode** [comment]
It looks like the proper place for this bug is at crbug. Thanks again @kylemacfarlane for doing all this work.

I'm going to close this one here. Please reopen if it turns out to be on this side.
